### ISSUE TITLE
go/build: honour GO15VENDOREXPERIMENT

### ISSUE THREAD

**Issue by vgalu** [golang/go#12278]
Hi,

Any chance of goimports being taught about GO15VENDOREXPERIMENT? With go 1.5 it rewrites import paths (to $CWD/vendor/path/to/3rd/party/pkg) even when GO15VENDOREXPERIMENT is set.

Thanks
Vlad

**Comment by bradfitz** [comment]
Yes, this needs to happen. Now that my projects are also using vendor directories, I've also been running into this.

/cc @adg @Sajmani @josharian

**Comment by ereyes01** [comment]
I also have issues with oracle and GO15VENDOREXPERIMENT and `oracle`, which I've initially reported here: fatih/vim-go#525

Is this correct place to raise an issue for oracle? If so should I open a new one, or report in this one?

Thanks, Eddy

**Comment by ianlancetaylor** [comment]
@ereyes01 Assuming you are talking about x/tools/cmd/oracle, then filing an issue on this issue tracker is the right thing to do.  This particular issue is about goimports.

**Comment by minux** [comment]
Shouldn't the functionality be integrated into go/build so
that every tool that uses go/build will automatically
handle the vendor experiment?

**Comment by kylewolfe** [comment]
@minux, this is what I've been trying to get my head around the last few days. Tools like goimports and gotools would need a significant refactor and very temporary workaround to implement support for the vendor experiment.

Is there a reason that the vendor experiment was implemented only in the go tool rather than in go/build?

**Comment by kylewolfe** [comment]
@rsc, Is it possible and even worth it at this point to move the vendor functionality from cmd/go to go/build to enable third party tools to piggy back on the existing vendor functionality?

**Comment by minux** [comment]
On Sep 4, 2015 11:29 AM, "Kyle Wolfe" <notifications@github.com> wrote:
> @rsc, Is it possible and even worth it at this point to move the vendor
functionality from cmd/go to go/build to enable third party tools to piggy
back on the existing vendor functionality?

One issue is that the new go/build package will only be available in Go
1.6. I don't know if that's acceptable.

**Comment by pwaller** [comment]
Could this be put onto a near-term milestone? (1.6?) I'm happily using the new vendoring mechanism and not having these source manipulation tools at our disposal is unfortunate. I'm also happy to have a crack at doing something if there is a consensus about what is a reasonable approach to fix it. Waiting until 1.6 is annoying, but waiting until *after* 1.6 would be worse.

**Comment by bradfitz** [comment]
Goimports is not distributed with Go so we don't use the Go 1.N milestones for it. This can be fixed whenever.

**Comment by pwaller** [comment]
@bradfitz I was referring to @minux's suggestion that the fix might involve fixing go/build. There are actually lots of things in need of fixing, and many of them are using go/build, so it is a logical thing that might need attention.

**Comment by pwaller** [comment]
For example, goimports uses go/build here: https://github.com/golang/tools/blob/5a90b2958f12ca0f046ffe2eb1523c5896af5dae/imports/fix.go#L158 - which implies either the stdlib needs fixing or the whole importing logic needs to move out of the stdlib.

**Comment by minux** [comment]
I modified this issue to track specifically the issue of go/build support
for the vendoring experiment.

As the vendoring experiment is going to be enabled by default in Go 1.6,
I think go/build support is also in order.

Label this Go1.6Early.

Once this is done, we can vendor the go/build package temporarily
into x/tools so that gorename/goimports could benefit Go 1.5 users.

**Comment by pwaller** [comment]
Ping.

I note that the 1.6Early deadline (according to the github milestone) is October the 31st, which is now ~12 days away. Is this issue on anybody's list, and is there anything I can do to help? Has the deadline already been missed, so to say? I would *really* like to be able to use `gorename` and friends in Go 1.6. It is already hurting.

**Comment by adg** [comment]
The deadline hasn't been missed.
But we should get moving on this.
If someone could investigate what's involved in making this work, that
would be great.

**Comment by cespare** [comment]
I started taking a look at this.

@minux, you wrote

> Shouldn't the functionality be integrated into go/build so
> that every tool that uses go/build will automatically
> handle the vendor experiment?

and you changed this bug from being about goimports to being about go/build. But there's no change to go/build that would make goimports just work with GO15VENDOREXPERIMENT, and in fact most or all the changes that would be required for goimports would be in x/tools/imports. (It could be made to work without any go/build changes, I think.)

So I suggest that either this issue be changed to be about goimports again, or we make a new bug to track the goimports changes.

I sent an email to golang-dev to solicit more ideas about go/build.

**Comment by abourget** [comment]
It's the 31st in some parts of the world already :)  I'd love to see that fixed!

**Comment by sp-adan-lobato** [comment]
I'd like also this issue to be fixed! :+1:

**Comment by gonzaloserrano** [comment]
Me too! Waiting to be able to pass all these awesome linters with the vendors' code instead of the global one.

**Comment by bradfitz** [comment]
Many people want this. No need to enumerate the list of them here.

**Comment by kylewolfe** [comment]
> I sent an email to golang-dev to solicit more ideas about go/build.

@cespare, can you link to discussion, please?

**Comment by cespare** [comment]
@kylewolfe my post is

https://groups.google.com/d/msg/golang-dev/kyckwHWpKQg/9nHp6kPzCwAJ

but it hasn't yielded any responses.

**Comment by kylewolfe** [comment]
Should we also discuss how internal is handled? Currently internal packages are resolved at cmd/go as well. Should this be handled in go/build? If not, are we ok with internal continuing to be resolved by cmd/go while vendor is handled by go/build?

**Comment by davecheney** [comment]
@kylewolfe vendor is also implemented in cmd/go, not go/build

https://go-review.googlesource.com/#/c/10923/

**Comment by kylewolfe** [comment]
@davecheney, Yes. The discussion is around moving vendor from cmd/go to go/build.

**Comment by davecheney** [comment]
@kylewolfe I'm sorry I misunderstood what you mean by "If not, are we ok with internal continuing to be resolved by cmd/go while vendor is handled by go/build?"

**Comment by cespare** [comment]
Please, this issue is already far too muddled without talking about internal packages as well. If you wish to discuss internal packages, please use the mailing list (or open a new issue if you feel there's a bug).

**Comment by kylewolfe** [comment]
@cespare, I'm sorry, but I think we should discuss where this leaves since it's doing almost exactly the same thing.

**Comment by dsymonds** [comment]
`internal` is easy to implement, and doesn't impact import paths. It doesn't need to be discussed on this issue, which is specifically about `vendor`.

**Comment by kylewolfe** [comment]
@davecheney, I could have phrased my entire post a little better :)

Here is the internal implementation review: https://codereview.appspot.com/120600043/

**Comment by FiloSottile** [comment]
Many tools now work out of the box just [rebuilding them with 1.6](https://blog.filippo.io/most-go-tools-now-work-with-go15vendorexperiment/), thanks!
